### PR TITLE
Doctor: Add dynamic check for incompatible Storybook packages

### LINKED ISSUE #26015: Fix doctor addon compatibility check
In https://github.com/storybookjs/storybook/issues/25293, we changed the CLI so that `sb upgrade` does not upgrade any non-core packages. The previous logic of “upgrade all the core packages + addons to the latest” wasn’t a terrible heuristic, but it also wasn’t great, and falls down when we could be upgrading to any arbitrary version and not the latest.

Since `sb upgrade` no longer upgrades addons, any user that uses non-core addons will very likely be left in a bad state after upgrade. To address this, we call `sb doctor` automatically after upgrades. `sb doctor`'s job is to look for inconsistencies in the user’s project setup, including addons that are incompatible with the current version of Storybook.

Unfortunately, the current logic for this is problematic. The list is hand-maintained, and hard-coded in the `doctor` command’s code, which means that (1) it is probably out of date, and (2) even when we update the list, it will probably not get patched to older versions of the CLI.

Instead of a hard-coding the list, we propose the following heuristics (or some variant thereof):
```
currentVersion = getCurrentVersion()

for addon in mainJs.addons:
  packageJson = readPackageJson(addon)
  for dep in storybookDeps(packageJson):
    if isIncompatible(dep, currentVersion):
      alert 'Addon ${addon}@${packageJson.version} likely incompatible'
      alert 'with Storybook ${currentVersion}.'
      alert 'Please see the addon docs: ${packageJson.homepage}'
      break
  latestVersion = getLatestVersion(addon)
  if(latestVersion > packageJson.version)
     alert 'There is a new version of ${addon} available: ${latestVersion}'
```

### PR DESCRIPTION
**Pull request by yannbf**
Closes #26015 

## What I did

This PR does the following:

- Removes the automigration for "incompatible-addons"
- Removes the previous logic for "incompatible-addons" check
- Adds a new logic to check incompatible Storybook dependencies (not just addons registered in main.js like before), and introduces it as part of the doctor command
- Makes the `automigrate` command run `doctor` at the end. This means the `upgrade` command will also run `doctor`
- Changes the logic so that the `getDuplicatedDepsWarnings` calls only happen if there are no incompatible packages,  reducing the noise considerably.

Here's how it looks like when running `doctor`:
image

Here's how it looks like when running `automigrate` or `upgrade`:
image

Here's how it looks like when running `storybook dev`:
image

### TODO:

1. This does not work with Yarn PnP
2. Check scenario with the canary when running `storybook dev`. There should be a check as well.
3. When developing this, I noticed two small issues in the package.json of our packages:
- ~@storybook/addon-coverage has a [main field](https://github.com/storybookjs/addon-coverage/blob/main/package.json#L19) but the file does not exist in dist~ (Done is [this PR](https://github.com/storybookjs/addon-coverage/pull/39))
- ~@storybook/addon-onboarding has a [require field](https://github.com/storybookjs/addon-onboarding/blob/main/package.json#L19) but the file does not exist in dist~ (Fixed in #26096)

We should really add an automated check with a tool like https://github.com/webpro/knip to automatically detect these issues

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Add an addon at a version that does not support the project's Storybook version, or just downgrade one of the packages to 7.0.0
4.1 Run `npx storybook doctor`
3.2  Run `npx storybook automigrate` (should yield a similar result)
3.3 Run `npx storybook upgrade` (should yield a similar result)
3.4 Run `storybook dev` (should yield a similar result, but faster as it does not check for available updates)

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This pull request has been released as version `0.0.0-pr-26219-sha-d52adbb1`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-26219-sha-d52adbb1 sandbox` or in an existing project with `npx storybook@0.0.0-pr-26219-sha-d52adbb1 upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-26219-sha-d52adbb1`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-26219-sha-d52adbb1) |
| **Triggered by** | @yannbf |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`yann/improve-incompatible-addon-check`](https://github.com/storybookjs/storybook/tree/yann/improve-incompatible-addon-check) |
| **Commit** | [`d52adbb1`](https://github.com/storybookjs/storybook/commit/d52adbb107d115c0dd1530038950cd213a315526) |
| **Datetime** | Wed Feb 28 11:59:09 UTC 2024 (`1709121549`) |
| **Workflow run** | [8079960307](https://github.com/storybookjs/storybook/actions/runs/8079960307) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=26219`_
</details>

### CODE REVIEW

File: code/lib/cli/src/automigrate/index.ts [hunk 1]
```diff
@@ -92,6 +93,8 @@ export const doAutomigrate = async (options: AutofixOptionsFromCLI) => {
     configDir,
     isUpgrade: false,
   });
+
+  await doctor({ configDir, packageManager: options.packageManager });
```
Comment (shilman) [inline code/lib/cli/src/automigrate/index.ts:97]: Don't we also need to remove one of the existing automigrations that overlaps with `doctor`?

**Review by shilman — COMMENTED** [review]
Should this be a draft PR or is this supposed to be mergeable?

File: code/lib/cli/src/automigrate/index.ts (hunk 1 above)
Comment (yannbf) [reply code/lib/cli/src/automigrate/index.ts:97]: yes, but that's already done

File: code/lib/cli/src/doctor/getIncompatibleStorybookPackages.test.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+      currentStorybookVersion: '8.0.0',
+      packageManager: packageManagerMock as JsPackageManager,
+    });
+    expect(result).toEqual({
+      packageName: 'my-storybook-package',
+      packageVersion: '1.0.0',
+      hasIncompatibleDependencies: true,
+      homepage: undefined,
+      availableUpdate: undefined,
+    });
```
Comment (ndelangen) [inline code/lib/cli/src/doctor/getIncompatibleStorybookPackages.test.ts:56]:
```suggestion
    expect(result).toEqual(expect.objectContaining({
      packageName: 'my-storybook-package',
      packageVersion: '1.0.0',
      hasIncompatibleDependencies: true,
    }));
```

Comment (ndelangen) [reply code/lib/cli/src/doctor/getIncompatibleStorybookPackages.test.ts:56]: This saves some lines, and makes it clear what's actually important.
Please apply similar to other tests.

File: code/lib/cli/src/doctor/getIncompatibleStorybookPackages.test.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    vi.mocked(doctorUtils.getPackageJsonOfDependency).mockResolvedValueOnce({
+      name: packageName,
+      version: '1.0.0',
+      dependencies: {
+        '@storybook/core-common': '7.0.0',
+      },
+    } as any);
+    const result = await checkPackageCompatibility(packageName, {
+      currentStorybookVersion: '8.0.0',
+      packageManager: packageManagerMock as JsPackageManager,
```
Comment (ndelangen) [inline code/lib/cli/src/doctor/getIncompatibleStorybookPackages.test.ts:48]:
```suggestion
      packageManager: packageManagerMock,
```

Can we make it accept the partial?

**Review by ndelangen — APPROVED** [review]

File: code/lib/cli/src/doctor/getIncompatibleStorybookPackages.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+export const getIncompatibleStorybookPackages = async (
+  context: Omit<Context, 'packageManager'> & Partial<Pick<Context, 'packageManager'>>
+): Promise<AnalysedPackage[]> => {
+  const packageManager = context.packageManager ?? JsPackageManagerFactory.getPackageManager();
+
+  const allDeps = await packageManager.getAllDependencies();
+  const storybookLikeDeps = Object.keys(allDeps).filter((dep) => dep.includes('storybook'));
+
+  if (storybookLikeDeps.length === 0) {
+    throw new Error('No storybook dependencies found in the package.json');
```
Comment (valentinpalkovic) [inline code/lib/cli/src/doctor/getIncompatibleStorybookPackages.ts:86]:
```suggestion
    throw new Error('No Storybook dependencies found in the package.json');
```

File: code/lib/cli/src/doctor/utils.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  if (!result?.packageJson) {
+    throw new PackageJsonNotFoundError();
+  }
+
+  return result.packageJson;
+};
+
+export const isPrerelease = (version: string) => {
+  return version.includes('-');
+};
```
Comment (valentinpalkovic) [inline code/lib/cli/src/doctor/utils.ts:35]: Can you elaborate on why you aren't using `JsPackageManager.getPackageJSON`, which is compatible with yarn pnp?

**Review by ndelangen — APPROVED** [review]

File: code/lib/cli/src/doctor/getIncompatibleStorybookPackages.ts [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+      dependencies,
+      peerDependencies,
+      homepage,
+    } = await getPackageJsonOfDependency(dependency);
+
+    const hasIncompatibleDependencies = !!Object.entries({
+      ...dependencies,
+      ...peerDependencies,
+    })
+      .filter(([dep]) => Object.keys(storybookCorePackages).includes(dep))
```
Comment (valentinpalkovic) [inline code/lib/cli/src/doctor/getIncompatibleStorybookPackages.ts:42]: Shouldn't this be enough? This saves a step transforming an object of key value pairs into a list of keys.

```suggestion
      .filter(([dep]) => storybookCorePackages[dep])
```

Even if you don't agree, we don't have to calculate `Object.keys(storybookCorePackages)` for every dependency. Let's move it one level up then and assign it to a variable.

File: code/lib/cli/src/doctor/getIncompatibleStorybookPackages.ts [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        const packageRepo = homepage ? `\n Repo: ${chalk.yellow(homepage)}` : '';
+
+        summaryMessage.push(`- ${packageDescription}${updateMessage}${packageRepo}`);
+      }
+    );
+
+    summaryMessage.push(
+      '\n',
+      'Please consider updating your packages or contacting the maintainers for compatibility details.',
+      'For more on Storybook 8 compatibility, see the linked Github issue:',
```
Comment (valentinpalkovic) [inline code/lib/cli/src/doctor/getIncompatibleStorybookPackages.ts:123]:
```suggestion
      'For more on Storybook 8 compatibility, see the linked GitHub issue:',
```

File: code/lib/cli/src/doctor/getIncompatibleStorybookPackages.test.ts [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    ];
+    const summary = getIncompatiblePackagesSummary(analysedPackages, '8.0.0');
+    expect(summary).toMatchInlineSnapshot(`
+      "The following packages are incompatible with Storybook 8.0.0 as they depend on different major versions of Storybook packages:
+      - storybook-react@1.0.0
+      - @storybook/addon-essentials@7.0.0 (8.0.0 available!)
+
+
+      Please consider updating your packages or contacting the maintainers for compatibility details.
+      For more on Storybook 8 compatibility, see the linked Github issue:
```
Comment (valentinpalkovic) [inline code/lib/cli/src/doctor/getIncompatibleStorybookPackages.test.ts:152]:
```suggestion
      For more on Storybook 8 compatibility, see the linked GitHub issue:
```

**Pull request merged**